### PR TITLE
Add deper validation of task definition YAML

### PR DESCRIPTION
**Pull request by joshma**
Uses jsonschema to (1) generate a schema from Definition{} and (2)
report on any errors.

This is meant as a lightweight validation layer to provide the end user
with better hints - if something isn't caught here, the server will
still error.

To support this, this introduces `utils.ErrorExplained` for better error
printing when we want to print out err.Error() followed by err.ExplainError()

Invalid YAML:

```
Error reading ../task-playground/create-validation/airplane.yml: invalid YAML

For more information on the task definition format, see the docs:
https://docs.airplane.dev/reference/task-definition-reference
```

Map instead of array:

```
Error reading ../task-playground/create-validation/airplane.yml

parameters: Invalid type. Expected: array, given: object

For more information on the task definition format, see the docs:
https://docs.airplane.dev/reference/task-definition-reference
```

Missing fields:

```
Error reading ../task-playground/create-validation/airplane.yml

parameters.0: type is required

For more information on the task definition format, see the docs:
https://docs.airplane.dev/reference/task-definition-reference
```

Invalid type:

```
Error reading ../task-playground/create-validation/airplane.yml

parameters.0.type: Invalid type. Expected: string, given: integer

For more information on the task definition format, see the docs:
https://docs.airplane.dev/reference/task-definition-referenc
```

Resolves AIR-607

### CODE REVIEW

**Comment by joshma** [comment]
On a minor note, @colinking wdyt of making the catch-all `Error: ...` red + unindented?

File: pkg/taskdir/taskdef.go [hunk 1]
```diff
@@ -56,6 +61,32 @@ func (this TaskDirectory) ReadDefinition() (Definition, error) {
...
+		} else if path, err := filepath.Rel(wd, defPath); err != nil {
+			logger.Debug("%s", err)
+		} else {
+			defPath = path
+		}
+		// Print any "expected" validation errors
+		switch err := errors.Cause(err).(type) {
+		case ErrInvalidYAML:
+			logger.Log(logger.Red("\nError reading %s: invalid YAML", defPath))
+			logger.Log("\nTask definition reference: %s", taskDefDocURL)
```
Comment (colinking) [inline pkg/taskdir/taskdef.go:79]: Maybe: `For more information on the task definition format, see the docs: %s`

File: pkg/taskdir/taskdef.go [hunk 2]
```diff
@@ -56,6 +61,32 @@ func (this TaskDirectory) ReadDefinition() (Definition, error) {
...
+		case ErrSchemaValidation:
+			logger.Log(logger.Red("\nError reading %s:\n", defPath))
+			for _, verr := range err.Errors {
+				logger.Log("  %s: %s", verr.Field(), verr.Description())
+			}
+			logger.Log("\nTask definition reference: %s", taskDefDocURL)
```
Comment (colinking) [inline pkg/taskdir/taskdef.go:85]: Maybe this logic live at the top-level (for converting an error into a log message)

File: pkg/taskdir/validate.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+// Returns ErrInvalidYAML if not valid YAML and ErrSchemaValidation if YAML doesn't match schemaObj
+func ValidateYAML(data []byte, schemaObj interface{}) error {
+	var obj interface{}
+	if err := yaml.Unmarshal(data, &obj); err != nil {
+		return errors.WithStack(ErrInvalidYAML{})
+	}
+
+	r := &jsonschema.Reflector{PreferYAMLSchema: true}
+	schemaLoader := gojsonschema.NewGoLoader(r.Reflect(schemaObj))
+	docLoader := gojsonschema.NewGoLoader(obj)
```
Comment (colinking) [inline pkg/taskdir/validate.go:34]: that's pretty cool

**Review by colinking — APPROVED** [review]

**Comment by joshma** [comment]
Heh I just realized `exit status 1` was coming from `go run` 😅

File: pkg/taskdir/taskdef.go (hunk 2 above)
Comment (colinking) [reply pkg/taskdir/taskdef.go:85]: (so you don't get two errors -- the error printed here, and the error returned to the top-level)

Comment (joshma) [reply pkg/taskdir/taskdef.go:85]: Ooh interesting - we could have some interface like `ExplainError()`?

**Review by francescaleung — APPROVED** [review]

**Comment by joshma** [comment]
Note the new changes! You can now implement [utils.ErrorExplained](https://github.com/airplanedev/cli/pull/101/files#diff-3dac6216516066e1c8e70c660b415e22307daf89d6e956baeb73cb318d1341d9R3) to have the CLI print out `err.Error()` followed by `err.ExplainError()`

**Pull request merged**